current/advanced/block-import.md

```diff
@@ -0,0 +1,41 @@
+---
+slug: block-import
+lang: en
+title: The Block Import Pipeline
+---
+
+
+## The Import Queue
+
+The import queue is an abstract worker queue present in every Substrate node. It is not part of the runtime. The Import Queue is responsible for processing pieces of incoming information, verifying them, and if they are valid, importing them into the node's state. The most fundamental piece of information that the import queue processes is blocks themselves, but it is also responsible for importing consensus-related messages such as Justifications and, in light clients, finality proofs.
+
+The Import queue collects incoming blocks from the network or from other parts of the local node, and stores them in a pool. The blocks are later checked for validity and discarded if they are not valid. Finally, the block is imported when the import queue is instructed to import it by the consensus engine.
+
+The import queue is codified abstractly in Substrate by means of the [`ImportQueue` trait](https://substrate.dev/rustdocs/v2.0.0-alpha.8/sp_consensus/import_queue/trait.ImportQueue.html). The use of a trait allows each consensus engine to provide its own specialized implementation of the Import Queue which may take advantage of optimization opportunities such as verifying multiple blocks in parallel as they come in across the network.
+
+## The Basic Queue
+Substrate also provides a default implementation known as the [`BasicQueue`](https://substrate.dev/rustdocs/v2.0.0-alpha.8/sp_consensus/import_queue/struct.BasicQueue.html). The `BasicQueue` does not do any kind of optimization, rather it performs the verification and import steps sequentially. It does, however, abstract the notion of verification through the use of the `Verifier`(https://substrate.dev/rustdocs/v2.0.0-alpha.8/sp_consensus/import_queue/trait.Verifier.html) trait.
+
+Any consensus engine that relies on the `BasicQueue` must implement the `Verifier` trait. In the simplest cases, such as Sunstrate's instant seal consensus, the `Verifier` is simple declares all blocks as valid. In slot-based engines check that the blocks timestamp actually falls in the correct time window for the slot it is claiming.
+
+## The Block Import Trait
+
+When the import queue is instructed to import a block, it passes the block in question to a method provided by the [`BlockImport` trait](https://substrate.dev/rustdocs/v2.0.0-alpha.8/sp_consensus/block_import/trait.BlockImport.html). This `Block Import` trait provides the behavior of importing a block into the node's local state.
+
+One implementor of the `BlockImport` trait that is used in every Substrate node is the [`Client`](https://substrate.dev/rustdocs/v2.0.0-alpha.8/sc_service/client/index.html) which contains the node's entire block database. When a block is imported into the client it is added to the main database of blocks that the node knows about.
```

Comment:
Seems like you're using "block database" and "node state" as synonyms; that is probably fine but some (most?) blockchains treat the blocks database separately from the state database. Maybe that's an uninteresting detail though. Either way: consider picking a single term for the final outcome of the import process.

Reply:
I was thinking of the block database as part of the node's state. If node state has a particular technical meaning, I did not know it and did not mean it. I just meant the state of the node. So in general when blocks get imported they update the node's state. When they get imported into the client specifically, they update the block database. But in a more complex import pipeline, they may update other pieces of local state (like the state of the grandpa voter for example) first, then ultimately be passed to the client where they update the block database.

TBH, what I'm explaining here is my current understanding. This whole article needs review from someone who actually knows the answers to these questions.